DRA: Introduce a helper function producing standardized 'resource.kubernetes.io/pcieRoot' device attributes for DRA drivers

everpeace opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR introduces a helper function `GetPCIeRootAttributeByPCIBusID(pciBusID)` for DRA drivers, allowing them to produce standardized device attributes in a consistent way. 

#### Which issue(s) this PR is related to:

ref: https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/401#discussion_r2144942856
KEP: https://github.com/kubernetes/enhancements/issues/4381

<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/4381

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduce a method 'GetPCIeRootAttributeByPCIBusID(pciBusID)' for third-party DRA drivers to provide common logic for the standardized device attribute 'resource.kubernetes.io/pcieRoot'
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4381-dra-structured-parameters
```

everpeace committed:
DRA: Introduce a helper function producing standardized device attributes for DRA drivers

The shared function can provide a way for DRA drivers to produce standardized device attributes in a consistent way.

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace commented:
/assign @aojea

michaelasp commented:
/cc @michaelasp

michaelasp commented on staging/src/k8s.io/dynamic-resource-allocation/utils/pci.go:
```diff
@@ -0,0 +1,213 @@
+	return fmt.Sprintf("%04x:%02x:%02x.%01x", p.Domain, p.Bus, p.Device, p.Function)
+}
+
+// ResolvePCIRoot resolves the PCI root for the PCIAddress.
```
I might be being ignorant here but why can't we just get the PCIRoot from the domain and bus of the pci address? If it's required to be done this way through the sysDevice path then I should do something similar for https://github.com/google/dranet/pull/114

michaelasp replied:
Thanks for the in depth explanation! I'm just starting to learn this from a more general networking background, makes sense to me :)

michaelasp commented on staging/src/k8s.io/dynamic-resource-allocation/utils/pci.go:
```diff
@@ -0,0 +1,213 @@
+// Example: "0000:0e:1f.0" represents domain 0, bus 14, device 31, function 0.
+//
+// ref: https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation
+type PCIAddress struct {
```
Should we ensure that address matches and does not exceed the size limits of the spec? I.e function can only be up to 15

everpeace replied:
Thanks for the info! Agreed. We should validate them. Would you know a pointer to official(-ish) spec? I'd like to know the specs for other parts(domain, bus, device)

michaelasp reviewed (commented):
This is awesome, thanks for adding this!

everpeace committed:
fix lint: apply De Morgan's laws to simplify conditionals

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace committed:
Validate PCI Address (domain is 16 bits, bus is 8 bits, device is 5 bits, function is 3 bits)

it also changes the data type from uint32 to uint16 for PCI Root and PCI Address

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace committed:
Add boilerplate header to go files

Signed-off-by: Shingo Omura <everpeace@gmail.com>

klueska commented:
I haven't reviewed this in detail yet, but at first glance it seems like there are multiple of layers of abstraction being introduced in this PR, for not much benefit.

When suggesting this addition in https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/401#discussion_r2144942856, I assumed we would just introduce a single package called `deviceattribute`, with a collection of methods that could be called against it to get a specific standard attribute.

With the very first method / set of constants being:
```
const (
    StandardAttributePrefix = "resource.kubernetes.io"
    StandardAttributePCIeRoot  resourceapi.QualifiedName = StandardAttributePrefix + "/pcieRoot"
)
func GetPCIeRootByBusID(busID string) resourceapi.DeviceAttribute
```

Where this would then be used as:
```
device.attributes[deviceattributes.StandardAttributePCIeRoot] = deviceattributes.GetPCIeRootByBusID(busID)
```

I could maybe be convinced that a top-level function that returns a map of attributes linking the attribute name to the attribute itself is useful but, I'm not convinced we should introduce three new folders with bits and pieces of functionality spread across them in order to enable this.

everpeace commented:
> I haven't reviewed this in detail yet, but at first glance it seems like there are multiple of layers of abstraction being introduced in this PR, for not much benefit.
> ...
> I could maybe be convinced that a top-level function that returns a map of attributes linking the attribute name to the attribute itself is useful but, I'm not convinced we should introduce three new folders with bits and pieces of functionality spread across them in order to enable this.


Thank you for your feedback, and sorry for overengineering. I'll update my branch so that `deviceattribute` will just provide `GetPCIeRootByBusID`, but, let me keep sysfs abstraction (I'll move it into `deviceattribute` package and make it private) for testing purpose.

everpeace committed:
Add StandardDeviceAttributes and StandardPCIDeviceAttributes helper functions

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace commented:
> > I could maybe be convinced that a top-level function that returns a map of attributes linking the attribute name to the attribute itself is useful
> 
> I'd like to keep this one for future scenario which we could add extra standardized attributes. Let me update the PR a bit more. So, I'll make this PR back to draft for a while.

I updated this PR:
- 5dd512b: simplified the implementation: `deviceattribute` package just provides `GetPCIeRootAttributeByPCIBusId` method (parent commit of 218efe7)
- 218efe7: `deviceattribute` package provides `StandardDeviceAttributes` method (current branch head)

PTAL and give me a feedback which we should choose 🙇

everpeace committed:
Simplify the implementation: just provide GetPCIeRootAttributeByPCIBusID

michaelasp commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci.go:
```diff
@@ -0,0 +1,149 @@
+	qualifiedNamePCIeRoot = resourceapi.QualifiedName(resourceapi.StandardDeviceAttributePCIeRoot)
+)
+
+type StandardPCIDeviceAttributesOption func(args *StandardPCIDeviceAttributeArgs)
+
+// WithPCIDeviceAddress sets the PCI address for the PCI Bus ID
+// in BDF (Bus-Device-Function) format, e.g., "0123:45:1e.7".
+//
+// It returns a DeviceAttribute with the PCIe Root Complex information("pci<domain>:<bus>")
+// as a string value or an error if the PCI Bus ID is invalid or the root complex cannot be determined.
+//
+// ref: https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation
+func WithPCIDeviceAddress(pciBusID string) StandardPCIDeviceAttributesOption {
+	return func(args *StandardPCIDeviceAttributeArgs) {
+		args.pciAddress = pciBusID
+	}
+}
+
+// StandardPCIDeviceAttributeArgs holds the arguments for generating standard PCI device attributes.
+type StandardPCIDeviceAttributeArgs struct {
+	pciAddress string
+	sysfs      sysfs
+}
+
+// StandardPCIDeviceAttributes returns standard device attributes for a PCI device.
+func StandardPCIDeviceAttributes(opts ...StandardPCIDeviceAttributesOption) (map[resourceapi.QualifiedName]resourceapi.DeviceAttribute, error) {
+	args := &StandardPCIDeviceAttributeArgs{
+		sysfs: sysfs(defaultSysfsRoot),
+	}
+	for _, opt := range opts {
+		opt(args)
+	}
+
+	attrs := make(map[resourceapi.QualifiedName]resourceapi.DeviceAttribute)
```
Nit: This could be a 1 line composite declaration on line 71 since we know the value off the bat.

```
attrs := map[resourceapi.QualifiedName]resourceapi.DeviceAttribute{qualifiedNamePCIeRoot: pciRootAttribute}
```

However since the types are long I can see breaking it up into two lines, but maybe we should move it below the error check?

klueska replied:
With my suggestion to have individual pkg level methods to grab the specific attribute of interest, I think this will be simplified significantly

michaelasp commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci.go:
```diff
@@ -0,0 +1,149 @@
+	return attrs, nil
+}
+
+// This is only for testing purposes.
```
Nit: Still should follow the convention for function comments, i.e

`withSysfs updates the sysfs path, this is only for testing purposes, the sysfs path will not change in a real world scenario`

michaelasp reviewed (commented):
Minor nits, otherwise LGTM!

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/api/standard_attributes.go:
We should drop this file / folder. This should just be in the deviceattribute folder itself.

everpeace replied:
fixed in [8ed8217](https://github.com/kubernetes/kubernetes/pull/132296/commits/8ed8217e9ab9a8f07677701202ef5fb80595e88a)

aojea commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/attribute.go:
```diff
@@ -0,0 +1,52 @@
+	pciDeviceAttributeOpts []StandardPCIDeviceAttributesOption
+}
+
+// StandardDeviceAttributes generates standard device attributes based on the provided options.
+// It currently supports standard PCI device attributes if options are provided.
+func StandardDeviceAttributes(
+	opts ...StandardDeviceAttributesOption,
+) (map[resourceapi.QualifiedName]resourceapi.DeviceAttribute, error) {
+	options := &StandardDeviceAttributesOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+	attrs := map[resourceapi.QualifiedName]resourceapi.DeviceAttribute{}
+	if len(options.pciDeviceAttributeOpts) > 0 {
+		// Standard PCI device attributes
+		pciAttrs, err := StandardPCIDeviceAttributes(options.pciDeviceAttributeOpts...)
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range pciAttrs {
+			attrs[k] = v
+		}
+	}
+	return attrs, nil
+}
```
agree, I think that since this is a library we should avoid to try to create abstractions to soon, because if we don't get it right we'll break consumers and people will just stop using it.

I think we should start small, let's expose the particular method for this attribute and once we get more feedback from drivers implementers we'll be able to come up with better

everpeace replied:
Thank you, guys. Then, let me update this PR so that it provides just `GetPCIRootByBusID` method.

everpeace committed:
Revert "Add StandardDeviceAttributes and StandardPCIDeviceAttributes helper functions"

This reverts commit 2869b6f42c4625bd72acea3cea77b639065b49c8.

everpeace committed:
Moved device attribute constants from api to deviceattribute package

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace committed:
Introduce DeviceAttribute{Name, Value} and change return type of 'GetPCIeRootAttributeByPCIBusID' with it

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace committed:
Move bdfRegexp from global constant to local variable

everpeace committed:
Put _linux suffix for linux specific go files

everpeace commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci.go:
fixed in [73d65f8](https://github.com/kubernetes/kubernetes/pull/132296/commits/73d65f830a1c8b095bee9433be4c318217f96d12)

everpeace committed:
merge constants.go into attribute.go

Signed-off-by: Shingo Omura <everpeace@gmail.com>

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go:
```diff
@@ -0,0 +1,97 @@
+	resourceapi "k8s.io/dynamic-resource-allocation/api"
+)
+
+// GetPCIeRootAttributeByPCIBusID retrieves the PCIe Root Complex for a given PCI Bus ID.
+// in BDF (Bus-Device-Function) format, e.g., "0123:45:1e.7".
+//
+// It returns a DeviceAttribute with the PCIe Root Complex information("pci<domain>:<bus>")
+// as a string value or an error if the PCI Bus ID is invalid or the root complex cannot be determined.
+//
+// ref: https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation
+func GetPCIeRootAttributeByPCIBusID(pciBusID string) (DeviceAttribute, error) {
```
I don't think I said this elsewhere, but the reason to be explicit about `ByPCIBusID` is that I think we will want a `GetPCIeRootAttributeByCpuID()` at some point soon.

klueska replied:
We want to map a given CPU to its PCIe root complex independent of NUMA. We've done alignment by NUMA in the past, but going forward direct alignment by PCIe root complex is the preferred approach.

One way to see this association on my 4-GPU system is as follows:
```
$ cat /sys/bus/pci/devices/0019:01:00.0/local_cpulist
72-143
$ cat /sys/bus/pci/devices/0018:01:00.0/local_cpulist
72-143
$ cat /sys/bus/pci/devices/0009:01:00.0/local_cpulist
0-71
$ cat /sys/bus/pci/devices/0008:01:00.0/local_cpulist
0-71
```
The first two GPUs are "local" to CPUs 72-143 and the second two GPUs are "local" to CPUs 0-71. Where "local" implies that they are on the same PCIe root complex. You can imagine how one could use this to implement a robust `GetPCIeRootAttributeByCpuID()` function.

There may be other more straightforward ways to glean this information, but this is gist of it.

/cc @kad

everpeace replied:
> The first two GPUs are "local" to CPUs 72-143, and the second two GPUs are "local" to CPUs 0-71. Where "local" implies that they are on the same PCIe root complex. You can imagine how one could use this to implement a robust GetPCIeRootAttributeByCpuID() function.

In my 8 GPUs system,

``` shell
$ nvidia-smi --query-gpu=gpu_bus_id --format=csv,noheader
00000000:04:00.0
00000000:23:00.0
00000000:43:00.0
00000000:64:00.0
00000000:84:00.0
00000000:A3:00.0
00000000:C3:00.0
00000000:E4:00.0
```

For example, For `0000:04:00.0`(1st) and `0000:64:00.0`(4th):
- both are "local" to `0-63`, 
- BUT, they have different PCIe roots
```
# For 0000:04:00.0:
#   its pcie root is 0000:00
$ readlink /sys/bus/pci/devices/0000:04:00.0
../../../devices/pci0000:00/0000:00:01.1/0000:01:00.0/0000:02:01.0/0000:04:00.0
#   local CPU is 0-63
$ cat /sys/bus/pci/devices/0000:04:00.0/local_cpulist
0-63

# For 0000:64:00.0:
#   its pcie root is 0000:60
$ readlink /sys/bus/pci/devices/0000:64:00.0
../../../devices/pci0000:60/0000:60:01.1/0000:61:00.0/0000:62:01.0/0000:64:00.0
#   local CPU is 0-63
$ cat /sys/bus/pci/devices/0000:64:00.0/local_cpulist
0-63
```

So, I think it would be difficult to determine CPU's "local" PCIe root. Are there any other way to determine/define it??

klueska replied:
Let's cross that bridge when it comes. I imagine we might prefer to introduce some other, single attribute to represent this. But we can decide that when the time is right.

johnbelamaric replied:
> which unfortunately won't work with simple matchAttributes

Depends how we define the semantics of `matchAttributes` if we add list typed attributes. We could define a "match" as a non-empty intersection between lists, and treat a scalar as a one-element list.

pohly commented:
> can you point us where this can be exercised in the e2e driver?

The test/e2e/dra/test-driver? That driver has no hardware dependency, so the code in this PR cannot be exercised through it.

In general, writing any kind of integration or E2E test for this in the Kubernetes CI is problematic because it needs some real, deterministic sysfs. We don't have that.

klueska commented:
Thanks @pohly that makes sense. Do you have any comments on the name of the package / API for this?

pohly commented:
/test pull-kubernetes-apidiff

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux_test.go:
```diff
@@ -0,0 +1,143 @@
+		expectedErrMsg    string
+	}{
+		"valid": {
+			smockSysfsSetup: func(t *testing.T, mockSysfs sysfsPath) {
```
What does the "s" in "smock" stand for?

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux_test.go:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+func TouchFile(t *testing.T, path string) {
```
```suggestion
func touchFile(t *testing.T, path string) {
```

It's easier to see that this is not a public API when it's not exported (doesn't make a different in _test.go, but still...).

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux_test.go:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+func CreateSymlink(t *testing.T, target, link string) {
```
```suggestion
func createSymlink(t *testing.T, target, link string) {
```

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/sysfs_linux.go:
```diff
@@ -0,0 +1,63 @@
+// and can be replaced with a mock in tests.
+type sysfsPath string
+
+func (s sysfsPath) Devices(path string) string {
```
```suggestion
func (s sysfsPath) devices(path string) string {
```

Same below.

In this case, exporting the method makes it callable outside of the package (assuming that some exported method returns a sysfsPath) to that caller and thus part of the package API. We want a lean API to make future changes safer.

everpeace committed:
Fix typo: s/smock/mock/g

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace committed:
Make test only helpers private (sysfspath methods, touchFile, createSymlink)

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace committed:
Commented out for unused sysfspath methods

Signed-off-by: Shingo Omura <everpeace@gmail.com>

everpeace reviewed (commented):
@pohly Thank you for your review. I addressed your comments. PTAL.

everpeace committed:
Remove unused sysfsPath's methods

Signed-off-by: Shingo Omura <everpeace@gmail.com>

klueska commented:
@kad please review the following thread when you get a chance: https://github.com/kubernetes/kubernetes/pull/132296#discussion_r2154600716

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go:
```diff
@@ -0,0 +1,97 @@
+		return DeviceAttribute{}, fmt.Errorf("failed to resolve sysfs path for PCI Bus ID %s: %w", pciBusID, err)
+	}
+
+	pciRootPart := strings.Split(strings.TrimPrefix(sysDevicesPath, sysfs.devices("")+"/"), "/")[0]
```
Does it make sense to ensure there are at least some number of parts to the split? Or is it expected to ever have just the root by the time we reach this stage?

everpeace replied:
> Does it make sense to ensure there are at least some number of parts to the split?
> Or is it expected to ever have just the root by the time we reach this stage?

Yes, I think so. 
`sysDevicesPath` is always of the form of `/sys/devices/pci0000:17/...`
ref: https://docs.kernel.org/PCI/sysfs-pci.html

everpeace replied:
New method `resolvePCIeRoot` method was introduced in https://github.com/kubernetes/kubernetes/pull/132296/commits/a4fb1562f4a50b1709ff0cf6263fc12f99a6d3e8

extracting pcieRoot part is performed just after sysDevicesPath validation. So, I think it's more readable now:
https://github.com/kubernetes/kubernetes/blob/872f0682d585c4929ae2b278d03fa4ab27e30892/staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go#L89-L101

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go:
```diff
@@ -0,0 +1,97 @@
+//
+// ref: https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation
+//
+// /sys/devices has directory structure which reflects the hardware hierarchy in the system.
+// Therefore, the device path may contains intermediate directories(devices).
+// Thus, we can not simply find the device path from the PCIAddress.
+// Fortunately, /sys/bus/pci/devices/<address> is a symlink to the actual device path in /sys/devices.
+// So we can resolve the actual device path by reading the symlink at /sys/bus/pci/devices/<address>.
+//
+// For example, if the PCIAddress is "0000:00:1f.0",
+// /sys/bus/pci/devices/0000:00:1f.0 points to
+// /sys/devices/pci0000:01/...<intermediate PCI devices>.../0000:00:1f.0,
```
```suggestion
// /sys/devices has a directory structure which reflects the hardware hierarchy in the system.
// Therefore, the device path may contain intermediate directories (devices).
// Thus, we can not simply find the device path from the PCIAddress.
// Fortunately, /sys/bus/pci/devices/<address> is a symlink to the actual device path in /sys/devices.
// So we can resolve the actual device path by reading the symlink at /sys/bus/pci/devices/<address>.
//
// For example, if the PCIAddress is "0000:00:1f.0",
// /sys/bus/pci/devices/0000:00:1f.0 points to
// /sys/devices/pci0000:01/...<intermediate PCI devices>.../0000:00:1f.0,
```

everpeace committed:
assign return value to a variable instead of using it directly

everpeace committed:
Update staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go

Co-authored-by: Kevin Klues <klueska@gmail.com>

everpeace committed:
introduce resolvePCIeRoot instead of resolveSysDevicesPath

everpeace commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go:
```diff
@@ -0,0 +1,97 @@
+		return DeviceAttribute{}, fmt.Errorf("invalid PCI Bus ID format: %s", pciBusID)
+	}
+
+	// e.g. /sys/devices/pci0000:01/...<intermediate PCI devices>.../0000:00:1f.0,
+	sysDevicesPath, err := resolveSysDevicesPath(pciBusID)
+	if err != nil {
+		return DeviceAttribute{}, fmt.Errorf("failed to resolve sysfs path for PCI Bus ID %s: %w", pciBusID, err)
+	}
+
+	pciRootPart := strings.Split(strings.TrimPrefix(sysDevicesPath, sysfs.devices("")+"/"), "/")[0]
```
Thanks, it makese sense. I'll fix it.

everpeace committed:
Update staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go

Co-authored-by: Kevin Klues <klueska@gmail.com>

everpeace committed:
Split the symlink target error cases into for each condition

everpeace committed:
Use "k8s.io/api/resource/v1beta2" instead of "k8s.io/dynamic-resource-allocation/api" for resource API imports.

everpeace commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go:
```diff
@@ -0,0 +1,97 @@
+	return DeviceAttribute{
+		Name:  StandardDeviceAttributePCIeRoot,
+		Value: resourceapi.DeviceAttribute{StringValue: &pciRootPart},
+	}, nil
```
fixed in f3d4b216f3a20a807d5ae258804c4f699c09ed50

everpeace commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go:
```diff
@@ -0,0 +1,97 @@
+	devicePathPrefix := sysfs.devices("pci")
+	if !strings.HasPrefix(targetAbs, devicePathPrefix) || filepath.Base(targetAbs) != pciBusID {
+		return "", fmt.Errorf("invalid symlink target for PCI Bus ID %s: %s", sysBusPath, targetAbs)
+	}
```
fixed in 31d2de59dd9ce112c599db7ebf1a17ecfc849e11

everpeace commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/attribute.go:
```diff
@@ -0,0 +1,41 @@
+package deviceattribute
+
+import (
+	resourceapi "k8s.io/dynamic-resource-allocation/api"
```
fixed in 872f0682d585c4929ae2b278d03fa4ab27e30892

everpeace reviewed (commented):
@klueska Sorry for my late/last-minute update. I think your comments are addressed. PTAL.

everpeace commented:
/retest

klueska commented:
Thanks for the back and forth on this. Let's get this merged now, with a followup to address the comment above, once the code bumping DRA to `v1` gets merged.

klueska commented:
/lgtm
/approve

bg-chun commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux_test.go:
```diff
@@ -0,0 +1,156 @@
+
+func TestGetPCIeRootBAttributeyPCIBusID(t *testing.T) {
+	pciBusID := "0000:01:02.3"
+	pcieRoot := "pci0000:01"
```
cc @everpeace 
Everything looks good, but how could the PCIe root of "0000:01:02.3" be "pci0000:01"?
It should be "pci0000:00".

bg-chun commented on staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go:
```diff
@@ -0,0 +1,102 @@
+// But, fortunately, /sys/bus/pci/devices/<address> is a symlink to the actual device path in /sys/devices.
+// So we can resolve the actual device path by reading the symlink at /sys/bus/pci/devices/<address>.
+//
+// For example, if the PCIAddress is "0000:00:1f.0",
+// /sys/bus/pci/devices/0000:00:1f.0 points to
+// /sys/devices/pci0000:01/...<intermediate PCI devices>.../0000:00:1f.0,
+// where "pci0000:01" is the PCIe Root.
```
cc @everpeace 
If the PCIAddress is "0000:00:1f.0",
/sys/bus/pci/devices/0000:00:1f.0 should points to
/sys/devices/pci0000:00/...<intermediate PCI devices>.../0000:00:1f.0,
where "pci0000:00" is the PCIe Root.


the bus 00 can not exist under the bus 01.

everpeace reviewed (commented):
@bg-chun Thank you for the catch! I will open a PR to fix your comments.

_updated: I opened a PR: https://github.com/kubernetes/kubernetes/pull/133198_
